ocaml/testsuite/tests/typing-layouts/rep_poly.ml

```diff
@@ -0,0 +1,475 @@
+|}]
+(* CR layouts v2.9: the default part is not quite correct *)
+
+(*************************************)
+(* Interaction with other attributes *)
+
+external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity" [@@unboxed]
+
+[%%expect{|
+Line 1, characters 37-39:
+1 | external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity" [@@unboxed]
+                                         ^^
+Error: Don't know how to unbox this type.
+       Only float, int32, int64, nativeint, and vector primitives can be unboxed.
+|}]
+
+external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity" [@@untagged]
+
+[%%expect{|
+Line 1, characters 37-39:
+1 | external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity" [@@untagged]
+                                         ^^
+Error: Don't know how to untag this type. Only int can be untagged.
+|}]
+
+(*************************************)
+(* Type var in nested in other types *)
+
+external[@rep_poly] id : ('a : any). 'a t_with_any -> 'a t_with_any = "%identity"
+
+let f (x: float#): float# = id x
+let f (x: int64#): int64# = id x
+let f (x: int32#): int32# = id x
+
+[%%expect{|
+external id : ('a : any). 'a t_with_any -> 'a t_with_any = "%identity"
+  [@@rep_poly]
+val f : float# -> float# = <fun>
+val f : int64# -> int64# = <fun>
+val f : int32# -> int32# = <fun>
+|}]
+
+
+external[@rep_poly] id : ('a : any). 'a M_any.t -> 'a M_any.t = "%identity"
+
+let f (): float# M_any.t = id (assert false : float# M_any.t)
+let f (): int64# M_any.t = id (assert false : int64# M_any.t)
+let f (): int32# M_any.t = id (assert false : int32# M_any.t)
+
+[%%expect{|
+external id : ('a : any). 'a M_any.t -> 'a M_any.t = "%identity" [@@rep_poly]
+val f : unit -> float# M_any.t = <fun>
+val f : unit -> int64# M_any.t = <fun>
+val f : unit -> int32# M_any.t = <fun>
+|}]
+
+(* Test this when sorts can be inside unboxed records *)
+(* type ('a : any) r = {field: 'a} [@@unboxed]
+external[@rep_poly] id : ('a : any). 'a M_any.t r -> 'a M_any.t r = "%identity"
+
+let f (): float# M_any.t r = id (assert false : float# M_any.t r)
+let f (): int64# M_any.t r = id (assert false : int64# M_any.t r)
+let f (): int32# M_any.t r = id (assert false : int32# M_any.t r) *)
+
+
+(********************************************)
+(* Some primitives require rep_poly to work *)
+
+type ('a : any) t
+external id : ('a : any). 'a t -> int = "%array_length"
+let id' x = id x
+
+[%%expect{|
+type ('a : any) t
+Line 2, characters 14-37:
+2 | external id : ('a : any). 'a t -> int = "%array_length"
+                  ^^^^^^^^^^^^^^^^^^^^^^^
+Error: The primitive [%array_length] doesn't work well with type variables of
+       layout any. Consider using [@rep_poly].
+|}]
+
+external[@rep_poly] id : ('a : any). 'a t -> int = "%array_length"
+let id' x = id x
+
+[%%expect{|
+external id : ('a : any). 'a t -> int = "%array_length" [@@rep_poly]
+val id' : 'a t -> int = <fun>
+|}]
+
+external id : ('a : any). 'a t -> int = "%identity"
```

Comment:
This one looks OK to me: the kind of `'a t` is `value`, regardless of the kind of `'a`. For `%array_length`, the situation is different, because that primitive assumes (a little strangely) that the argument looks like `'a array` for some representable `'a`. `%identity` does not do this.

Reply:
Yes, I misread this one - I agree it doesn't need to be an error.  But as noted elsewhere I'm strugging to come up with an a good reason to want `('a : any) . ...` without `rep_poly`, and think we should just ban it anyway.

Reply:
Maybe that's true for primitives, but it's easier to imagine C functions that would want this behavior. That is, you can imagine a C function that processes some type with a phantom parameter. Allowing the external really to have an `any` variable would then allow further abstraction over the externals. If we require `rep_poly`, you couldn't propagate the polymorphism. Without looking deeply at the code, it surprises me that this is hard to arrange for. Perhaps when I next catch up with Alan, we'll look at this. (If it is indeed hard to keep the polymorphism, then let's not do the heavy lifting until we need to.)

Reply:
Sure, that sounds reasonable.  I think concretely what I'd suggest for the long term is  to ban `('a : any) . ...` on builtin externals, unless they are marked `rep_poly` and on a list of builtins we know to work with `rep_poly`.  I think this is compatible with your goal.

In the short term `('a : any) . ...` simply doesn't do what it sounds like it does, so I think it's reasonable to extend the above to ban it on all non-`rep_poly` externals, not just builtins, but I don't feel strongly.

To keep the polymorphism and achieve your goal we'd have to change the tyvar `jkind_initialization` policy for externals from `Sort` to `Any`.  Probably this could be done, you and Alan will know better than me why the current policy was chosen.

Reply:
Agreed with the above.